Add rate limit detection

stefanvermaas opened the pull request:
Fixes #23 

This adds rate limit detection to the Fulfil gem. Allowing clients to watch for the `Fulfil::RateLimitExceeded` exception and handle accordingly. 

Also exposes helpful information about the current quota:
- `Fulfil.rate_limit.requests_left?` (aliased by `Fulfil.rate_limit.exceeded?`)
- `Fulfil.rate_limit.limit` exposes the maximum number of requests you're permitted to make per second.
- `Fulfil.rate_limit.resets_at` exposes the time at which the current rate limit window resets in UTC epoch seconds.

cdmwebs commented:
> what if the headers aren't set?

This seems unlikely. I would expect the library to assume that the request was allowed and proceed. The rate limit helpers should be a progressive enhancement.

> This is more of a problem for test suites

You mentioned that in Slack. It doesn't seem like it would affect many suites at first glance.

👍🏻

cdmwebs commented on lib/fulfil/rate_limit.rb:
```diff
@@ -0,0 +1,57 @@
+      @resets_at = nil
+    end
+
+    # Analyses the rate limit based on the response headers from Fulfil.
```
Love the documentation. I'll have to get in to this habit 😄

stefanvermaas replied:
For gems, I actually like adding documentation. Gems are often a bit more abstract than application code. So, I figured it made sense to add it here too.

swanny85 replied:
It definitely helps someone new to the code get a better footing.

cdmwebs commented on lib/fulfil/response_handler.rb:
```diff
@@ -43,8 +54,6 @@ def verify!
       )
     end
 
-    private
```
Was this intentional?

stefanvermaas commented:
@cdmwebs I didn't want the test suites - even though not a lot of people use this gem YET ;) - not to fail. So, I've extracted the parsing of the HTTP response headers into a new class and added some default values to it. I think this is the best and friendliest way to handle this.

If you're okay with this, I'll make sure to merge it and to release a new version.

cdmwebs commented:
@stefanvermaas makes sense! :shipit:
